src/main/scala/subsystem/HasTiles.scala

```diff
@@ -86,7 +108,68 @@ trait HasTileInterruptSources
   }
 }
 
+/** These are sources of "constants" that are driven into the tile.
+  * 
+  * While they are not expected to change dyanmically while the tile is executing code,
+  * they may be either tied to a contant value or programmed during boot or reset.
+  * They need to be instantiated before tiles are attached within the subsystem containing them.
+  */
+trait HasTileInputConstants extends InstantiatesTiles { this: BaseSubsystem =>
+  /** tileHartIdNode is used to collect publishers and subscribers of hartids. */
+  val tileHartIdNode = BundleBridgeEphemeralNode[UInt]()
+
+  /** tileHartIdNexusNode is a BundleBridgeNexus that collects dynamic hart prefixes,
+    *   orReduces them, and combines the reduction with the static ids assigned to each tile,
+    *   producing a unique, dynamic hartid for each tile.
+    */
+  val tileHartIdNexusNode = BundleBridgeNexus[UInt](
+    inputFn = BundleBridgeNexus.orReduction[UInt](registered = p(HartPrefixKey)) _,
+    outputFn = (prefix: UInt, n: Int) =>  Seq.tabulate(n) { i =>
+      val y = dontTouch(prefix | hartIdList(i).U(p(MaxHartIdBits).W))
+      if (p(HartPrefixKey)) BundleBridgeNexus.safeRegNext(y) else y
+    },
+    default = Some(() => 0.U(p(MaxHartIdBits).W)),
+    inputRequiresOutput = true // guard against this being driven but ignored in tileHartIdIONodes below
+  )
+  // TODO: Replace the DebugModuleHartSelFuncs config key with logic to consume the dynamic hart IDs
+
+  /** tileResetVectorNode is used to collect publishers and subscribers of tile reset vector addresses. */
+  val tileResetVectorNode = BundleBridgeEphemeralNode[UInt]()
+
+  /** tileResetVectorNexusNode is a BundleBridgeNexus that accepts a single reset vector source, and broadcasts it to all tiles. */
+  val tileResetVectorNexusNode = BundleBroadcast[UInt](
+    inputRequiresOutput = true // guard against this being driven but ignored in tileResetVectorIONodes below
+  )
+
+  /** tileHartIdIONodes may generate subsystem IOs, one per tile, allowing the parent to assign unique hart ids.
+    *
+    *   Or, if such IOs are not configured to exist, tileHartIdNexusNode is used to supply an id to each tile.
+    */
+  val tileHartIdIONodes: Seq[BundleBridgeSource[UInt]] = p(SubsystemExternalHartIdWidthKey) match {
```

Comment:
I am still struggling with the relationships between p(MaxHartIdWidth) and p(SubsystemExternalHartIdWidthKey). 

I believe that the former is the width that will be used in $mhartid CSR. So should p(SubsystemExternalHartIdWidthKey) (if defined) be <= p(MaxHartIdBits)? Or is it allowed to be smaller?

Reply:
IMO its allowed to be smaller.

Reply:
Ok, and if it were larger, the truncation check would catch it, so I think this is OK

Reply:
After further digging around in the closed source side, i found this:
`case MaxHartIdBits => 10 // maximum possible before debug unit requirements fail`
which implies to me that not only can it be smaller it in fact MUST be smaller.

This does make me wonder whether I should switch the core IOs over to just using `XLen` though, rather than have this somewhat extraneous parameter floating around in core-parameter-land just to make a layer or two of still-narrow IOs that then get put into an `XLen`-sized CSR?

Reply:
that might be a question for @aswaterman

Reply:
Oh actually it looks like the physical CSR is actually sized by the width of this IO wire as well, so i will leave the `hartIdLen` parameter present to control that.